### ISSUE TITLE
Web Workers: Bundle shared dependencies only once

### ISSUE THREAD

**Issue by jase88** [angular/angular-cli#16001]
# 🚀 feature request

### Relevant Package
This feature request is for @angular-devkit/build-angular

### Description
If you currently add a web worker to Angular (e.g. via `ng generate webWorker`) and use the same dependency on the main and the worker thread, it is bundled twice.

For example with a dependency like moment:

**app.worker.ts**
```
import * as moment from 'moment';

addEventListener('message', ({ data }) => {
  const response = `worker response to ${data} at ${moment().format()}`;
  postMessage(response);
});

```
**app.component.ts**
```
import * as moment from 'moment';
```

This can have radical impacts on your bundle size! It would be great if every dependency would only bundled once. The same occurs if using multiple workers with the same dependency.

Code splitting like this would improve the usage of web workers with Angular.

//edit: also indirectly discussed here: https://github.com/angular/angular-cli/issues/15033 

### Describe the solution you'd like
Automatically detect shared dependencies and let webpack generate a separate file for that dependency. So the main bundle could import it as well as the web worker via *importScripts*.

### Describe alternatives you've considered
Manual way to isolate a dependency to a separate file on build (configuration). This would at least enable us to import that file manually via *importScripts* on a web worker.

**Comment by filipesilva** [comment]
I understand the motivation but I'm not 100% sure this is feasible at the moment. The way https://github.com/GoogleChromeLabs/worker-plugin (the way we make the web worker) works is by creating a whole separate sub compilation. That compilation is general shares nothing with the main one. I suppose the first step is to have a feature request on `worker-plugin`.

**Comment by kerosin** [comment]
Okay. I'm not familiar with Webpack compilation and what is happening under the hood.

What would the ticket on the worker plugin be about? What needs to be done?

I guess there have to be things done on multiple ends.

**Comment by trotyl** [comment]
> Okay. I'm not familiar with Webpack compilation and what is happening under the hood.
> 
> What would the ticket on the worker plugin be about? What needs to be done?
> 
> I guess there have to be things done on multiple ends.

It doesn't seem to be feasible at all.

1. Worker build doesn't share the same `tsconfig`, the build result can be different;
2. Chunks are loaded via JSONP, which cannot be done in worker;
3. Splitting can only happen in lazy-loading, not static imports;

**Comment by kerosin** [comment]
Relevant https://github.com/webpack/webpack/issues/6472

**Comment by onzag** [comment]
@kerosin https://www.npmjs.com/package/worker-injector-generator-plugin

Mixed everyone's knowledge on the matters and created this simple fix, it works.

Relevant question https://stackoverflow.com/questions/49354186/webpack-common-chunks-for-code-shared-between-webworker-and-web-code

And also the thread you passed, were all sources.

I am by no means a webpack expert just mixed up what everyone was saying in that plugin.

**Comment by alan-agius4** [comment]
Thanks for reporting this issue. Luckily, it has already been fixed in one of the recent releases. Please update to the most recent version to resolve the problem.

If the problem persists in your application after upgrading, please open a new issue, provide a simple repository reproducing the problem, and describe the difference between the expected and current behavior. You can use `ng new repro-app` to create a new project where you reproduce the problem.
